From 95b03ac1ccb06ba3d18b83cbb6f1ce058daa7418 Mon Sep 17 00:00:00 2001
From: Luiz Angelo Daros de Luca <luizluca@gmail.com>
Date: Wed, 11 Jan 2023 02:08:02 -0300
Subject: [PATCH 06/12] bitfield.h: add FIELD_WIDTH()

Some hardware stores logical fields split across multiple registers,
requiring partial values to be reassembled using shifts and masks.

Add FIELD_WIDTH(), a helper that returns the number of contiguous bits
covered by a bitfield mask. This allows callers to determine how much a
partial value needs to be shifted when reconstructing a field.

This complements FIELD_MAX() and FIELD_FIT(), and avoids reimplementing
bit counting logic in drivers.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>
---
 include/linux/bitfield.h | 13 +++++++++++++
 1 file changed, 13 insertions(+)

--- a/include/linux/bitfield.h
+++ b/include/linux/bitfield.h
@@ -90,6 +90,19 @@
 	})
 
 /**
+ * FIELD_WIDTH() - return the width of a bitfield
+ * @_mask: shifted mask defining the field's length and position
+ *
+ * Returns the number of contiguous bits covered by @_mask.
+ * This corresponds to the bit width of FIELD_MAX(@_mask).
+ */
+#define FIELD_WIDTH(_mask)						\
+	({								\
+		__BF_FIELD_CHECK(_mask, 0ULL, 0ULL, "FIELD_WIDTH: ");	\
+		__bf_shf(~FIELD_MAX(_mask));				\
+	})
+
+/**
  * FIELD_FIT() - check if value fits in the field
  * @_mask: shifted mask defining the field's length and position
  * @_val:  value to test against the field
